### ISSUE TITLE
No summary when run with streaming

### ISSUE THREAD

**Issue by kedashoe** [scottcorgan/tap-spec#37]
When I run

`node test/test.js | node_modules/.bin/tap-spec`

I get the nice

```
total:     1
passing:   0
failing:   1
duration:  49ms
```

output at the end, but when I set up streaming

```
Tape.createStream().pipe(TapSpec()).pipe(process.stdout);
```

and run

```
node test/test.js
```

there is no summary at the end. Currently running `iojs-v2.1.0`. Figure things are just closing before parser gets the output event? Not sure if I am missing something simple or if there is an easy fix. If not I am happy to look into it more. Thanks for the great package!

**Comment by scottcorgan** [comment]
Can you post your full code so I can duplicate no my end?

**Comment by kedashoe** [comment]
Sure, thank you for the quick reply.

sample-pipe.js
```js
var Tape = require('tape');

Tape('hello', function(t){
    t.assert(1, 'world');
    t.end();
});
```
output
```
$ node test/sample-pipe.js | node_modules/.bin/tap-spec 

  hello

    ✓ world


  total:     1
  passing:   1
  duration:  13ms
```
sample-stream.js
```js
var Tape = require('tape');
var TapSpec = require('tap-spec');

Tape.createStream()
  .pipe(TapSpec())
  .pipe(process.stdout);

Tape('hello', function(t){
    t.assert(1, 'world');
    t.end();
});
```
output
```
$ node test/sample-stream.js 

  hello

    ✓ world
```
```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="14.04.2 LTS, Trusty Tahr"
$ node --version
v2.1.0
```

**Comment by angrbrd** [comment]
I'm curious if any further work has been done on this issue? I'm running into the same thing. Thanks!

**Comment by jmm** [comment]
Same thing here. [Minimal repro](https://github.com/jmm/scottcorgan-tap-spec-37-repro). Environment:

Ubuntu 12.04
node@5.6.0

**Comment by rctyler** [comment]
I'm also running into this problem, any work-arounds?

**Comment by scottcorgan** [comment]
I am open for PR. I won't have time for a few more weeks to get into this. Anyone up for it?

**Comment by erinspice** [comment]
I just saw this behavior tonight.  I'm writing a test suite where each test is dependent upon a unique and correct `process.argv` value. As such, each test is a separate process and streams its stdout to tap-spec. I think I found a solution. Let me write it up properly.

**Comment by erinspice** [comment]
Well, I was going to, until I cloned the repo and found there are no tests at all. Here's an example in which streaming returns the summary, in case anyone cares.

Clone [this repo](https://github.com/erinspice/quasix-getopt/commit/38e60a91866e150c1af6613450c05d34a3ca9619), `npm i`, and run the tests with `npm test`.

**Comment by antonycourtney** [comment]
**TL;DR:** A workaround is to explicitly instantiate a test harness.  [Example code](https://github.com/jmm/scottcorgan-tap-spec-37-repro/compare/master...antonycourtney:master)

I was also getting bit by this, so I rolled up my sleeves and did some digging, first in tap-spec, and then in `tap-out` and then in `tape` itself.  I think I've found the cause and a (clumsy) workaround.

To my eyes it appears that tape itself will, by default, only properly `close()` the output stream if the process exits.  It achieves this by registering a handler for the process `'exit'` event in the `exitHarness` that it creates by default.  See: https://github.com/substack/tape/blob/master/index.js#L75-L91

A (somewhat awkward) workaround is to explicitly create a distinct test harness instance using `test.createHarness()`.  This returns a new test harness that can be used just like `test(...)`, but must be passed around explicitly.  Harnesses created via this call go through a different code path ([`createHarness`](https://github.com/substack/tape/blob/master/index.js#L103-L151) vs [`createExitHarness`](https://github.com/substack/tape/blob/master/index.js#L55-L94)) that closes the output stream when all tests complete rather than when the process exits.  Unfortunately there is no mechanism I could find to direct tape to just use `createHarness` instead of `createExitHarness` by default.

I've modified the minimal repro from @scottcorgan to demonstrate how to call `createHarness` and use it explicitly for running tests.  Diffs are here:
https://github.com/jmm/scottcorgan-tap-spec-37-repro/compare/master...antonycourtney:master

Hope that's helpful.

**Comment by scottcorgan** [comment]
I'm going to close this, although we can continue our conversation. It's been a year and half and I'm not sure if this is still an issue or y'all have moved on.
